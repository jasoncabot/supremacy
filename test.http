### Signup
POST http://localhost:5173/api/auth/signup
Content-Type: application/json
X-Client-Id: test-client-id

{
  "username": "testuser",
  "password": "testpassword",
  "email": "test@example.com"
}

### Login
# @name login

POST http://localhost:5173/api/auth/login
Content-Type: application/json
X-Client-Id: test-client-id

{
  "username": "testuser",
  "password": "testpassword"
}

### Mint new refresh token
@refreshToken = {{login.response.body.refreshToken}}

POST http://localhost:5173/api/auth/refresh
Content-Type: application/json
X-Client-Id: test-client-id

{
  "refreshToken": "{{refreshToken}}"
}

### Create a game
POST http://localhost:5173/api/games
Content-Type: application/json
Authorization: Bearer {{login.response.body.accessToken}}
X-Client-Id: test-client-id

{
	"faction": "Empire",
	"difficulty": "Easy",
	"galaxySize": "Small",
	"winCondition": "Standard",
	"mode": "Single Player"
}
