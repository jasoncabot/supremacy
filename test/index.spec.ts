import { SELF } from "cloudflare:test";
import { afterAll, beforeAll, describe, expect, it } from "vitest";
import type {
	CreateGameRequest,
	LoginRequest,
	RefreshTokenRequest,
	SignupRequest,
} from "../worker/api";
import { makeDeterministic } from "./determinism";

// Standard headers that should be included in all API requests
// This follows the same pattern as in useApi.ts to ensure consistency
const standardHeaders = {
	"Content-Type": "application/json",
	"X-Client-ID": "test-client-id",
};

describe("api", () => {
	it("returns appropriate content for version", async () => {
		const request = new Request("http://example.com/api/version");
		const response = await SELF.fetch(request);
		expect(await response.json()).toMatchSnapshot();
	});

	it("returns appropriate content for health", async () => {
		const request = new Request("http://example.com/api/health");
		const response = await SELF.fetch(request);
		expect(await response.json()).toMatchSnapshot();
	});

	it("returns appropriate content for something-that-is-random", async () => {
		const request = new Request(
			"http://example.com/api/something-that-is-random",
		);
		const response = await SELF.fetch(request);
		expect(await response.json()).toMatchSnapshot();
	});
});

describe("auth", () => {
	// Store original functions to restore them later
	let originalRandomUUID: typeof crypto.randomUUID;
	let originalGetRandomValues: typeof crypto.getRandomValues;
	let originalDateNow: typeof Date.now;
	let originalMathRandom: typeof Math.random;

	beforeAll(() => {
		// Store original functions
		originalRandomUUID = crypto.randomUUID;
		originalGetRandomValues = crypto.getRandomValues;
		originalDateNow = Date.now;
		originalMathRandom = Math.random;

		makeDeterministic();
	});

	afterAll(() => {
		// Restore original functions
		crypto.randomUUID = originalRandomUUID;
		crypto.getRandomValues = originalGetRandomValues;
		Date.now = originalDateNow;
		Math.random = originalMathRandom;
	});

	it("can run through the whole sign up, login and refresh flow", async () => {
		const signupRequest = new Request("http://example.com/api/auth/signup", {
			method: "POST",
			body: JSON.stringify({
				username: "test",
				password: "test",
			} as SignupRequest),
			headers: standardHeaders,
		});
		const signupResponse = await SELF.fetch(signupRequest);
		expect(await signupResponse.json()).toMatchSnapshot();

		const loginRequest = new Request("http://example.com/api/auth/login", {
			method: "POST",
			body: JSON.stringify({
				username: "test",
				password: "test",
			} as LoginRequest),
			headers: standardHeaders,
		});
		const loginResponse = await SELF.fetch(loginRequest);
		expect(await loginResponse.json()).toMatchSnapshot();

		const refreshRequest = new Request("http://example.com/api/auth/refresh", {
			method: "POST",
			body: JSON.stringify({
				refreshToken:
					"swr:9cc45b3f05af7c62c2f9024c47b9e48f38b8160c2b04a5cc5acb8521775ed38d:11111111111111111111111111111111",
			} as RefreshTokenRequest),
			headers: standardHeaders,
		});
		const refreshResponse = await SELF.fetch(refreshRequest);
		expect(await refreshResponse.json()).toMatchSnapshot();

		// create a game with this access token
		const gameRequest = new Request("http://example.com/api/games", {
			method: "POST",
			body: JSON.stringify({
				faction: "Empire",
				difficulty: "Easy",
				galaxySize: "Small",
				winCondition: "Standard",
				mode: "Single Player",
			} as CreateGameRequest),
			headers: {
				...standardHeaders,
				Authorization:
					"Bearer swa:9cc45b3f05af7c62c2f9024c47b9e48f38b8160c2b04a5cc5acb8521775ed38d:22222222222222222222222222222222",
			},
		});
		const gameResponse = await SELF.fetch(gameRequest);
		expect(await gameResponse.json()).toMatchSnapshot();
	});
});
