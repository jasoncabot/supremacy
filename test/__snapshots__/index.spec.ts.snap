// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`api > returns appropriate content for health 1`] = `
{
  "status": "ok",
}
`;

exports[`api > returns appropriate content for something-that-is-random 1`] = `
{
  "error": "Not Found",
  "status": 404,
}
`;

exports[`api > returns appropriate content for version 1`] = `
{
  "version": "0.1.0",
}
`;

exports[`auth > can run through the whole sign up, login and refresh flow 1`] = `
{
  "accessToken": "swa:9cc45b3f05af7c62c2f9024c47b9e48f38b8160c2b04a5cc5acb8521775ed38d:22222222222222222222222222222222",
  "accessTokenExpiry": 1746403200000,
  "clientId": "test-client-id",
  "refreshToken": "swr:9cc45b3f05af7c62c2f9024c47b9e48f38b8160c2b04a5cc5acb8521775ed38d:33333333333333333333333333333333",
  "refreshTokenExpiry": 1748908800000,
}
`;

exports[`auth > can run through the whole sign up, login and refresh flow 2`] = `
{
  "accessToken": "swa:9cc45b3f05af7c62c2f9024c47b9e48f38b8160c2b04a5cc5acb8521775ed38d:44444444444444444444444444444444",
  "accessTokenExpiry": 1746403200000,
  "clientId": "test-client-id",
  "refreshToken": "swr:9cc45b3f05af7c62c2f9024c47b9e48f38b8160c2b04a5cc5acb8521775ed38d:11111111111111111111111111111111",
  "refreshTokenExpiry": 1748908800000,
}
`;

exports[`auth > can run through the whole sign up, login and refresh flow 3`] = `
{
  "accessToken": "swa:9cc45b3f05af7c62c2f9024c47b9e48f38b8160c2b04a5cc5acb8521775ed38d:22222222222222222222222222222222",
  "accessTokenExpiry": 1746403200000,
  "clientId": "test-client-id",
  "refreshToken": "swr:9cc45b3f05af7c62c2f9024c47b9e48f38b8160c2b04a5cc5acb8521775ed38d:33333333333333333333333333333333",
  "refreshTokenExpiry": 1748908800000,
}
`;

exports[`auth > can run through the whole sign up, login and refresh flow 4`] = `
{
  "gameId": "game-4",
}
`;
